### PR TITLE
[Do not merge/review] chore(ci): Add connectivity integration test for standalone instance COMPASS-4577

### PR DESCRIPTION
**Pull request by Anemy**
COMPASS-4577

This PR adds a connectivity integration test that tests the `connection-model`, `compass-connect`, and `data-service` together. It spins up a localhost docker instance, renders the connection form connecting string input with enzyme, connects using the connect button, and then verifies the connection details produced by the data service match what we expect. 

The `compass-connect` rendering and connecting in this pr only tests the connection model which the connection string generates to connect. We'll another test for the connection form connect button in another pr/ticket.

Hopefully evergreen picks up the test and runs it 🤞

### CODE REVIEW

File: test/connectivity/docker-instance-manager.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+const dockerEnvsRepoCommit = '7297467d208a8bfc15d4f53faf983ba9cbe87360';
+
+const directoryExists = (directoryPath) => {
+  // eslint-disable-next-line no-sync
+  return fs.existsSync(directoryPath) && fs.lstatSync(directoryPath).isDirectory();
+};
+
+const cloneDockerEnvsRepo = () => {
+  if (directoryExists(dockerEnvsPath)) {
+    execSync(`rm -Rf ${dockerEnvsPath}`);
```
Comment (addaleax) [inline test/connectivity/docker-instance-manager.js:16]: I always get all tingly when I see this :) I don’t know if anybody will ever try to run this in a path with spaces or other shell-relevant characters, but can we maybe add quotes around the path to make this a little more foolproof (mostly thinking about people who run this test locally here)?

```suggestion
    execSync(`rm -Rf "${dockerEnvsPath}"`);
```

(For the other `execSync()` calls as well, I guess, but `rm -Rf` is the one where it matters :slightly_smiling_face:)

**Review by addaleax — APPROVED** [review]
nice :)

File: test/connectivity/docker-instance-manager.js (hunk 1 above)
Comment (Anemy) [reply test/connectivity/docker-instance-manager.js:16]: Good call - I think that rm rf might not really be needed either, I'll see if we can do it another way just to ensure if someone changes up that `dockerEnvsPath` it doesn't delete something important.

**Comment by Anemy** [comment]
Seeing if we can update our ubuntu versions - right now we're on 1404 and would be nice to be 1804

File: test/connectivity/connectivity.test.jsx [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    _id: 'localhost:27020',
+    hostname: 'localhost',
+    port: 27020,
+    client: {
+      isWritable: true,
+      isMongos: false,
+    },
+    build: {
+      raw: {
+        version: /4.4.[1-9]+$/
```
Comment (Anemy) [inline test/connectivity/connectivity.test.jsx:106]: This version check will error when there's a minor version update - I'll update to only check for major updates 🤔 or maybe any version that is a valid version

File: test/connectivity/connectivity.test.jsx [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+          throw dataServiceConnectedErr;
+        }
+
+        if (timesConnectedCalled > 1) {
+          throw new Error('data-service-connected called multiple times');
+        }
+
+        const dataService = compassConnectStore.dataService;
+
+        // 4. Fetch the instance details using the new connection.
```
Comment (mcasimir) [inline test/connectivity/connectivity.test.jsx:218]: i so think it would be clearer to have different `it()` for all of those.

File: test/connectivity/connectivity.test.jsx [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+const { promisify } = require('util');
+
+Enzyme.configure({ adapter: new Adapter() });
+const { mount } = Enzyme;
+
+// Stub electron module for tests
+const originalLoader = m._load;
+const stubs = {
+  electron: {
+    ipcMain: {
```
Comment (mcasimir) [inline test/connectivity/connectivity.test.jsx:17]: why is this necessary?

File: test/connectivity/connectivity.test.jsx [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+      };
+      global.hadronApp.appRegistry = appRegistry;
+      global.hadronApp.appRegistry.registerRole('Application.Status', ROLE);
+
+      compassConnectStore = appRegistry.getStore('Connect.Store');
+
+      // Remove all logic around saving and loading stored connections.
+      // NOTE: This is tightly coupled with the store in compass-connect.
+      // https://github.com/mongodb-js/compass-connect/blob/master/src/stores/index.js
+      compassConnectStore._saveRecent = () => {};
```
Comment (mcasimir) [inline test/connectivity/connectivity.test.jsx:147]: what happens if we just let this happen? wouldn't this just work?

**Comment by mcasimir** [comment]
Super cool stuff Rhys!

File: test/connectivity/connectivity.test.jsx [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+        expect(instanceDetails.genuineMongoDB).to.deep.equal(
+          expectedInstanceDetails.genuineMongoDB
+        );
+        expect(instanceDetails.hostname).to.equal(
+          expectedInstanceDetails.hostname
+        );
+        expect(instanceDetails.port).to.equal(
+          expectedInstanceDetails.port
+        );
+      });
```
[the remark this answers is not in the document]
Comment (Anemy) [reply test/connectivity/connectivity.test.jsx:256]: ah good call, adding.

**Review by mcasimir — APPROVED** [review]

File: test/connectivity/connectivity.test.jsx [hunk 7]
```diff
@@ -0,0 +1,259 @@
+const { expect } = require('chai');
```
[the remark this answers is not in the document]
Comment (Anemy) [reply test/connectivity/connectivity.test.jsx]: Updated to `.spec.js`. TIL there's a difference lol

File: test/connectivity/connectivity.test.jsx [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+        // Here we use the parsed connection model and build a url.
+        // This is something that would occur if
+        // a user is switching between the connection views and editing.
+        compassConnectStore.state.customUrl = compassConnectStore.state.currentConnection.driverUrlWithSsh;
+        compassConnectStore.trigger(compassConnectStore.state);
+
+        const wrapper = mount(<CompassConnectComponent />);
+
+        // Ensures the model doesn't cause any errors when rendering.
+        await compassConnectStore.validateConnectionString();
```
[the remark this answers is not in the document]
Comment (Anemy) [reply test/connectivity/connectivity.test.jsx]: Updated - looks a like cleaner now, thanks

File: test/connectivity/connectivity.test.jsx [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+// Hide react warnings.
+const originalWarn = console.warn.bind(console.warn);
+console.warn = (msg) => (
+  !msg.toString().includes('componentWillReceiveProps')
+  && !msg.toString().includes('componentWillUpdate')
+  && originalWarn(msg)
+);
+
+const delay = promisify(setTimeout);
+const ensureConnected = async(timeout, testIsConnected) => {
```
[the remark this answers is not in the document]
Comment (Anemy) [reply test/connectivity/connectivity.test.jsx]: updated 👍

File: test/connectivity/connectivity.test.jsx (hunk 4 above)
Comment (Anemy) [reply test/connectivity/connectivity.test.jsx]: I tried using `electron-mocha` to fully mock the electron functions, however the `hadron-ipc` module that `storage-mixin` requires I think depends either on an older version of electron's commands, or requires both the renderer and main process to be running. I spent a bit of time trying to get it work and couldn't quite get it, so I figured this does the job. Totally open to suggestions here, I think it might be nice to keep these tests lightweight so I was avoiding spectron.

These are two of the electron parts we need to mock for reference:
https://github.com/mongodb-js/hadron-ipc/blob/master/lib/renderer.js#L10
https://github.com/mongodb-js/compass-connect/blob/master/src/components/form/form-file-input.jsx#L9

Comment (mcasimir) [reply test/connectivity/connectivity.test.jsx]: Was the issue something like 'the module did not register'?

File: test/connectivity/connectivity.test.jsx (hunk 5 above)
Comment (mcasimir) [reply test/connectivity/connectivity.test.jsx]: I mean, avoiding mocking and stubbing at this level would make those tests less of the "white-box" type, which i feel would make it more stable.

The main concern I have is that those mocks are relying on internal details of things that are defined quite far from here. Would not expect to break a test about 'compass can connect with tls' if I refactor the way recent connections are stored.

File: test/connectivity/connectivity.spec.js [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+      }
+    },
+    client: {
+      isWritable: true,
+      isMongos: false,
+    },
+    dataLake: { isDataLake: false, version: null },
+    genuineMongoDB: { isGenuine: true, dbType: 'mongodb' },
+    host: {
+      os: 'Ubuntu',
```
Comment (mcasimir) [inline test/connectivity/connectivity.spec.js:98]: doesn't this break locally?

File: test/connectivity/connectivity.spec.js [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+      expect(compassConnectStore.state.syntaxErrorMessage).to.equal(null);
+      expect(compassConnectStore.state.isValid).to.equal(true);
+    });
+
+    context(`Connection ${testConnectionTitle} successfully connects`, () => {
+      let dataService;
+      let instanceDetails;
+      let onDataServiceConnected;
+
+      before(async() => {
```
Comment (mcasimir) [inline test/connectivity/connectivity.spec.js:189]: What is the reason to not put this in a `beforeEach` instead? Is better if to avoid interaction between executions, the cost in terms of time should not be that big.

Comment (mcasimir) [reply test/connectivity/connectivity.spec.js:189]: Oh i see the 'it' are just checking stuff on the data we fetch in the before. I guess is ok.

File: test/connectivity/connectivity.spec.js [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+      };
+      global.hadronApp.appRegistry = appRegistry;
+      global.hadronApp.appRegistry.registerRole('Application.Status', ROLE);
+
+      compassConnectStore = appRegistry.getStore('Connect.Store');
+
+      // Remove all logic around saving and loading stored connections.
+      // NOTE: This is tightly coupled with the store in compass-connect.
+      // https://github.com/mongodb-js/compass-connect/blob/master/src/stores/index.js
+      compassConnectStore._saveRecent = () => {};
```
Comment (mcasimir) [inline test/connectivity/connectivity.spec.js:142]: I'd still put all of this mocking somewhere else.

File: test/connectivity/connectivity.spec.js [hunk 13]
```diff
@@ -0,0 +1,305 @@
...
+      // Here we use the parsed connection model and build a url.
+      // This is something that would occur if
+      // a user is switching between the connection views and editing.
+      compassConnectStore.state.customUrl = compassConnectStore.state.currentConnection.driverUrlWithSsh;
+      compassConnectStore.trigger(compassConnectStore.state);
+
+      wrapper = mount(<CompassConnectComponent />);
+
+      // Ensures the model doesn't cause any errors when rendering.
+      await compassConnectStore.validateConnectionString();
```
Comment (mcasimir) [inline test/connectivity/connectivity.spec.js:169]: isn't this some kind of `it` as well?

File: test/connectivity/connectivity.test.jsx (hunk 4 above)
Comment (Anemy) [reply test/connectivity/connectivity.test.jsx]: It was cannot access `dialog` of undefined https://github.com/mongodb-js/compass-connect/blob/master/src/components/form/form-file-input.jsx#L9
or cannot call undefined function call https://github.com/mongodb-js/hadron-ipc/blob/master/lib/renderer.js#L10 (which happens when running in the electron main process and I think that makes sense because that's expected to be run in the renderer process)

File: test/connectivity/connectivity.spec.js (hunk 10 above)
Comment (Anemy) [reply test/connectivity/connectivity.spec.js:98]: That's the host of the docker image, which should be ubuntu. Tests pass locally on my mac 👍

File: test/connectivity/connectivity.spec.js (hunk 11 above)
Comment (Anemy) [reply test/connectivity/connectivity.spec.js:189]: I think it will make sense to move them to beforeEach if they're doing extra things on the connection, right now I think it doesn't totally make sense to reconnect for each of these tests. The cost might get higher for external connections, since it would be synchronously reconnecting/disconnecting and sometimes running a query. If it's decent latency could add up quick, but yeah not too much. I'm thinking let's move them into before each if they need guaranteed isolation. Maybe it makes sense to make them beforeEach if we want to be fully testing the disconnect ability of compass as well 🤔

File: test/connectivity/connectivity.spec.js (hunk 13 above)
Comment (Anemy) [reply test/connectivity/connectivity.spec.js:169]: Ah yup that can be in the `it` below which is testing the side effects of this method on the store 😅 nice catch

File: test/connectivity/connectivity.spec.js (hunk 10 above)
Comment (Anemy) [reply test/connectivity/connectivity.spec.js:98]: Thinking about that though - hopefully we can get one of our tests running on a host machine that isn't ubuntu for test sake.

File: test/connectivity/connectivity.spec.js (hunk 12 above)
Comment (Anemy) [reply test/connectivity/connectivity.spec.js:142]: Seeing how we can avoid stubbing them altogether

File: test/connectivity/connectivity.test.jsx (hunk 5 above)
Comment (Anemy) [reply test/connectivity/connectivity.test.jsx]: Without stubbing this `saveRecent_` function we save the resulting connection into a folder called `Connections` at the root of Compass. I think it'll take a bit more work to either clean up this folder post run, or even better I think we can use this saved connection as a way to improve our test (by ensuring it saves the connection in the correct format, and maybe even loads it in afterwords). If it sounds good to you I'd like to break this into another ticket so we can merge what we have here and keep working on some other connections, since I think that should be more of an isolated change.
Here's a ticket for it I added to this project:  COMPASS-4630

**Comment by Anemy** [comment]
@mcasimir added a comment re: the mocking of saving, and cleaned up how we do some of the setup. Ready for another look!

**Comment by Anemy** [comment]
Chatted with Maurizio, we're thinking we'll mono repo before following up on this PR. Going to leave it open with a do not merge/review tag. Moving those tickets back into open.

**Comment by Anemy** [comment]
Closing this since we're adding end to end tests with typescript in the new epic.
